Library/Homebrew/cmd/pull.rb

```diff
@@ -217,4 +231,141 @@ def pull
       end
     end
   end
+
+  private
+
+  def pull_patch(url)
+    PatchPuller.new(url).pull_patch
+  end
+
+  class PatchPuller
+    attr_reader :base_url
+    attr_reader :patch_url
+    attr_reader :patchpath
+
+    def initialize(url)
+      @base_url = url
+      # GitHub provides commits/pull-requests raw patches using this URL.
+      @patch_url = url + ".patch"
+      @patchpath = HOMEBREW_CACHE + File.basename(patch_url)
+    end
+
+    def pull_patch
+      fetch_patch
+      apply_patch
+    end
+
+    def fetch_patch
+      ohai "Fetching patch"
+      curl patch_url, "-o", patchpath
+    end
+
+    def apply_patch
+      # Applies a patch previously downloaded with fetch_patch()
+      # Deletes the patch file as a side effect, regardless of success
+
+      ohai "Applying patch"
+      patch_args = []
+      # Normally we don't want whitespace errors, but squashing them can break
+      # patches so an option is provided to skip this step.
+      if ARGV.include?("--ignore-whitespace") || ARGV.include?("--clean")
+        patch_args << "--whitespace=nowarn"
+      else
+        patch_args << "--whitespace=fix"
+      end
+
+      # Fall back to three-way merge if patch does not apply cleanly
+      patch_args << "-3"
+      patch_args << patchpath
+
+      begin
+        safe_system "git", "am", *patch_args
+      rescue ErrorDuringExecution
+        if ARGV.include? "--resolve"
+          odie "Patch failed to apply: try to resolve it."
+        else
+          system "git", "am", "--abort"
+          odie "Patch failed to apply: aborted."
+        end
+      ensure
+        patchpath.unlink
+      end
+    end
+  end
+
+  # List files changed by a patch, partitioned in to those that are (probably)
+  # formula definitions, and those which aren't. Only applies to patches on 
+  # Homebrew core or taps, based simply on relative pathnames of affected files.
+  def files_changed_in_patch(patchfile)
+    files = []
+    formulae = []
+    others = []
+    File.foreach(patchfile) do |line|
+      files << $1 if line =~ /^\+\+\+ b\/(.*)/
+    end
+    files.each do |file|
+      if file.start_with?("Library/Formula") || file.match("^[^/]*\.rb")
```

Comment:
>  since that looks like it returns the full path to the tap's formula directory as installed on disk, where this method is looking at a patch file with pathnames relative to the repo root

That's why I use `tap.path` which returns the root path of the tap.

The reason I prefer to use such abstraction is we support many places to put formula(i.e. root dir, `Formula` and `HomebrewFormula`). And we will drop `Library/Formula` after core formulae separation. So it's better to center all these logic inside `Tap#formula_dir` method. In this way, you won't need to worry the implement detail beneath it.